Create nn_modules_sources structure, map params and placeholders

voznesenskym opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #95831
* __->__ #95825

This PR Introduces a new structure on the tracing context, a ModuleContext, with a mapping of name to source.

The original motivation is to (A) move dedup guards off positional arguments structures (B) Add support for dedup guards on params.

To that end, we lack a structure that is:

1) Not positional
2) Can record both guards and placeholders
3) Can help us get a source out for a given input or parameter

This PR introduces the structure mentioned above. The idea here is that, we can use this to keep a mapping of all param names, buffer names, and module names, as well as placeholder names, to their sources. Then, anywhere within the tracing context that has access to any of the above can always get their full, proper, source, and therefore, their name in user code. This is particularly important for creating guards in a system downstream of dynamo, where we do no have VariableTracker. 

cc @soumith @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

voznesenskym committed:
Add normalize_attr_name util

[ghstack-poisoned]

voznesenskym committed:
Update on "Add normalize_attr_name util"




cc soumith yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

voznesenskym committed:
Update on "Add normalize_attr_name util"




cc soumith yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

wconstab reviewed (approved):
i think lgtm. 

i noticed it is not strictly code refactor, you also added .lower() -is that ok?

jansel commented on torch/_dynamo/utils.py:
```diff
@@ -1371,3 +1371,11 @@ def tensor_shape_should_be_static(
+    name = re.sub(r"\[(\d+)\]", r"_\g<1>", name)
+    # e.g. replace abc.xyz_123.qkv with abc_xyz_123_qkv
+    name = re.sub(r"[^a-zA-Z0-9]", "_", name)
+    return name.lower()
```
Why lower?

voznesenskym replied:
Later on, we use this to normalize lookups, and there were cases where placeholder names and attr names were case mismatched.

ezyang replied:
I'm skeptical. In the context here, there's no requirement that names have to be consistent against calls, because we still uniqueify it below. You seem to imply that you actually need similar things to match.

ezyang reviewed (changes requested):
While there is nothing technically wrong with this refactor, the code is misused in future PRs. More comments coming.

ezyang commented:
https://github.com/pytorch/pytorch/pull/95826/files#r1123171042

voznesenskym committed:
Update on "Add normalize_attr_name util"




cc soumith yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

voznesenskym commented on torch/_dynamo/variables/builder.py:
```diff
@@ -584,15 +584,15 @@ def wrap_sym(self, value: Union[torch.SymInt, torch.SymFloat]):
             return self.tx.output.register_attr_or_module(
                 value,
-                re.sub(r"[^a-zA-Z0-9]+", "_", self.name),
+                normalize_attr_name(self.name),
```
Sadly, we need a little extra normalization here, so we can have a consistent lookup.

ezyang replied:
This doesn't make sense to me. You'll have to explain more. Alt, in a pinch, I can check out your PR and take a look at the error.

voznesenskym replied:
Yes, will add comments here to explain. Do not worry about pulling it down, its not ready yet, and not a good use of your time!

voznesenskym commented on torch/_dynamo/utils.py:
```diff
@@ -1371,3 +1373,11 @@ def tensor_shape_should_be_static(
+    name = re.sub(r"\[(\d+)\]", r"_\g<1>", name)
+    # e.g. replace abc.xyz_123.qkv with abc_xyz_123_qkv
+    name = re.sub(r"[^a-zA-Z0-9]", "_", name)
+    return name
```
Note - we should not need lower here

voznesenskym commented:
CI failures are real, ugh

voznesenskym committed:
Update on "Create nn_modules_sources structure, map params and placeholders"




cc soumith yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

voznesenskym committed:
Update on "Create nn_modules_sources structure, map params and placeholders"




cc soumith yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

ezyang commented on torch/_guards.py:
```diff
@@ -341,6 +363,15 @@ def current_frame(frame_summary):
         finally:
             tc.frame_summary_stack.pop()
 
+    def register(self, name: str, source: Source):
```
What's "name" here?

voznesenskym committed:
Update on "Create nn_modules_sources structure, map params and placeholders"




cc soumith yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

voznesenskym committed:
Update on "Create nn_modules_sources structure, map params and placeholders"


This PR Introduces a new structure on the tracing context, a ModuleContext, with a mapping of name to source.

The original motivation is to (A) move dedup guards off positional arguments structures (B) Add support for dedup guards on params.

To that end, we lack a structure that is:

1) Not positional
2) Can record both guards and placeholders
3) Can help us get a source out for a given input or parameter

This PR introduces the structure mentioned above. The idea here is that, we can use this to keep a mapping of all param names, buffer names, and module names, as well as placeholder names, to their sources. Then, anywhere within the tracing context that has access to any of the above can always get their full, proper, source, and therefore, their name in user code. This is particularly important for creating guards in a system downstream of dynamo, where we do no have VariableTracker. 

cc soumith yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

voznesenskym committed:
Update on "Create nn_modules_sources structure, map params and placeholders"


This PR Introduces a new structure on the tracing context, a ModuleContext, with a mapping of name to source.

The original motivation is to (A) move dedup guards off positional arguments structures (B) Add support for dedup guards on params.

To that end, we lack a structure that is:

1) Not positional
2) Can record both guards and placeholders
3) Can help us get a source out for a given input or parameter

This PR introduces the structure mentioned above. The idea here is that, we can use this to keep a mapping of all param names, buffer names, and module names, as well as placeholder names, to their sources. Then, anywhere within the tracing context that has access to any of the above can always get their full, proper, source, and therefore, their name in user code. This is particularly important for creating guards in a system downstream of dynamo, where we do no have VariableTracker. 

cc soumith yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

voznesenskym committed:
Update on "Create nn_modules_sources structure, map params and placeholders"


This PR Introduces a new structure on the tracing context, a ModuleContext, with a mapping of name to source.

The original motivation is to (A) move dedup guards off positional arguments structures (B) Add support for dedup guards on params.

To that end, we lack a structure that is:

1) Not positional
2) Can record both guards and placeholders
3) Can help us get a source out for a given input or parameter

This PR introduces the structure mentioned above. The idea here is that, we can use this to keep a mapping of all param names, buffer names, and module names, as well as placeholder names, to their sources. Then, anywhere within the tracing context that has access to any of the above can always get their full, proper, source, and therefore, their name in user code. This is particularly important for creating guards in a system downstream of dynamo, where we do no have VariableTracker. 

cc soumith yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

voznesenskym committed:
Update on "Create nn_modules_sources structure, map params and placeholders"


This PR Introduces a new structure on the tracing context, a ModuleContext, with a mapping of name to source.

The original motivation is to (A) move dedup guards off positional arguments structures (B) Add support for dedup guards on params.

To that end, we lack a structure that is:

1) Not positional
2) Can record both guards and placeholders
3) Can help us get a source out for a given input or parameter

This PR introduces the structure mentioned above. The idea here is that, we can use this to keep a mapping of all param names, buffer names, and module names, as well as placeholder names, to their sources. Then, anywhere within the tracing context that has access to any of the above can always get their full, proper, source, and therefore, their name in user code. This is particularly important for creating guards in a system downstream of dynamo, where we do no have VariableTracker. 

cc soumith yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

voznesenskym commented on torch/_dynamo/output_graph.py:
```diff
@@ -441,24 +470,51 @@ def wrap_name(module_key):
+                if isinstance(target, torch.nn.Module) and not is_lazy_module(target):
+                    # annoying, but there are cases when we do not have parameters
+                    # see test_nn_moduledict_contains
+                    if hasattr(target, "_parameters"):
+                        for n, p in target.named_parameters(recurse=False, remove_duplicate=False):
+                            new_source = ParamBufferSource(source, n)
+                            new_name = new_source.name()
+                            self.register_attr_or_module(p, new_name, source=new_source)
+                    # annoying, but there are cases when we do not have buffers
+                    # see test_nn_moduledict_contains
+                    if hasattr(target, "_buffers"):
+                        for n, p in target.named_buffers(recurse=False, remove_duplicate=False):
+                            new_source = ParamBufferSource(source, n)
+                            new_name = new_source.name()
+                            self.register_attr_or_module(p, new_name, source=new_source)
```
Reviewers - do you prefer we share the 3 common lines here, or leave it like this for readability?

voznesenskym committed:
Update on "Create nn_modules_sources structure, map params and placeholders"


This PR Introduces a new structure on the tracing context, a ModuleContext, with a mapping of name to source.

The original motivation is to (A) move dedup guards off positional arguments structures (B) Add support for dedup guards on params.

To that end, we lack a structure that is:

1) Not positional
2) Can record both guards and placeholders
3) Can help us get a source out for a given input or parameter

This PR introduces the structure mentioned above. The idea here is that, we can use this to keep a mapping of all param names, buffer names, and module names, as well as placeholder names, to their sources. Then, anywhere within the tracing context that has access to any of the above can always get their full, proper, source, and therefore, their name in user code. This is particularly important for creating guards in a system downstream of dynamo, where we do no have VariableTracker. 

cc soumith yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

voznesenskym commented on torch/_dynamo/output_graph.py:
```diff
@@ -441,24 +476,55 @@ def wrap_name(module_key):
-        # e.g. replace abc.xyz_123.qkv with abc_xyz_123_qkv
-        name = re.sub(r"[^a-zA-Z0-9]", "_", name)
 
+        # NOTE! This code used to strip square brackets,
+        # but we no longer do so in favor of a single uniformly used
+        # normalize_attr_name. We used to need this because of things we did with module
+        # lookups where we would eval with a name, and the square brackets would throw it off.
+        # We no longer do this.
+        #
+        # The reason we need this uniform is that we need all the names to be the same, so that
+        # all the sources can be the same, because when we eventually call .register into the
+        # context_module, we need to make sure elements match.
```
I think I need to rewrite this for clarity, it is wordy without saying a lot.

voznesenskym committed:
Update on "Create nn_modules_sources structure, map params and placeholders"


This PR Introduces a new structure on the tracing context, a ModuleContext, with a mapping of name to source.

The original motivation is to (A) move dedup guards off positional arguments structures (B) Add support for dedup guards on params.

To that end, we lack a structure that is:

1) Not positional
2) Can record both guards and placeholders
3) Can help us get a source out for a given input or parameter

This PR introduces the structure mentioned above. The idea here is that, we can use this to keep a mapping of all param names, buffer names, and module names, as well as placeholder names, to their sources. Then, anywhere within the tracing context that has access to any of the above can always get their full, proper, source, and therefore, their name in user code. This is particularly important for creating guards in a system downstream of dynamo, where we do no have VariableTracker. 

cc soumith yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

ezyang commented on torch/_guards.py:
```diff
@@ -269,6 +290,31 @@ def __eq__(self, other):
+
+    Returns None if no delta is found, otherwise, return a set() of mismatched
+    attr and parameter names.
+    """
```
put this docblock in diff itself?

ezyang commented on torch/_guards.py:
```diff
@@ -290,6 +336,41 @@ def restore_graphstate(self, state):
         self.dynamo_guards = state.dynamo_guards
 
 
+class ModuleContext(Checkpointable[ModuleCheckpointState]):
```
docs?

ezyang commented on torch/_guards.py:
```diff
@@ -290,6 +336,41 @@ def restore_graphstate(self, state):
+
+    def register(self, source: Source):
+        """
+        Registers a source object to it's given name.
```
but no name is given?

ezyang commented on torch/_guards.py:
```diff
@@ -290,6 +336,41 @@ def restore_graphstate(self, state):
+        :param source: A Source object representing the source to be registered.
+        :type source: Source
+
+        :raises AssertionError: If a source with the same name already exists but has a different source name.
```
This still seems misdesigned. If I'm reading this correctly, this structure tells you what the source of any named buffers/parameters on the GraphModule being built are. There should never be conflicts, because for every distinct source we should have explicitly allocated a fresh name for them.

voznesenskym replied:
There are never conflicts, and we never try catch this anywhere, but I wanted a way of detecting potentially illegal state.

Let me make this far more clear.

ezyang replied:
Normalize_attr_name is not a mangler. You can't guarantee distinct sources won't conflict

voznesenskym replied:
Understood. You said this before, but it did not click till now.

What if we had normalize_attr_name take the tracing context's name_to_sources, and collisions would always increment? I think that might work, right? Let me try a version of this PR that does that.

ezyang replied:
This still sounds structurally wrong. *Somewhere* you're doing name allocation, right? Wherever that is, that should be the one source of truth of names, and then you plumb the name along to wherever you need it otherwise.

voznesenskym replied:
Name for eval and name for lookup cannot be the same. So, no, we cannot just have one name.

voznesenskym committed:
Update on "Create nn_modules_sources structure, map params and placeholders"


This PR Introduces a new structure on the tracing context, a ModuleContext, with a mapping of name to source.

The original motivation is to (A) move dedup guards off positional arguments structures (B) Add support for dedup guards on params.

To that end, we lack a structure that is:

1) Not positional
2) Can record both guards and placeholders
3) Can help us get a source out for a given input or parameter

This PR introduces the structure mentioned above. The idea here is that, we can use this to keep a mapping of all param names, buffer names, and module names, as well as placeholder names, to their sources. Then, anywhere within the tracing context that has access to any of the above can always get their full, proper, source, and therefore, their name in user code. This is particularly important for creating guards in a system downstream of dynamo, where we do no have VariableTracker. 

cc soumith yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

voznesenskym committed:
Update on "Create nn_modules_sources structure, map params and placeholders"


This PR Introduces a new structure on the tracing context, a ModuleContext, with a mapping of name to source.

The original motivation is to (A) move dedup guards off positional arguments structures (B) Add support for dedup guards on params.

To that end, we lack a structure that is:

1) Not positional
2) Can record both guards and placeholders
3) Can help us get a source out for a given input or parameter

This PR introduces the structure mentioned above. The idea here is that, we can use this to keep a mapping of all param names, buffer names, and module names, as well as placeholder names, to their sources. Then, anywhere within the tracing context that has access to any of the above can always get their full, proper, source, and therefore, their name in user code. This is particularly important for creating guards in a system downstream of dynamo, where we do no have VariableTracker. 

cc soumith yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

voznesenskym committed:
Update on "Create nn_modules_sources structure, map params and placeholders"


This PR Introduces a new structure on the tracing context, a ModuleContext, with a mapping of name to source.

The original motivation is to (A) move dedup guards off positional arguments structures (B) Add support for dedup guards on params.

To that end, we lack a structure that is:

1) Not positional
2) Can record both guards and placeholders
3) Can help us get a source out for a given input or parameter

This PR introduces the structure mentioned above. The idea here is that, we can use this to keep a mapping of all param names, buffer names, and module names, as well as placeholder names, to their sources. Then, anywhere within the tracing context that has access to any of the above can always get their full, proper, source, and therefore, their name in user code. This is particularly important for creating guards in a system downstream of dynamo, where we do no have VariableTracker. 

cc soumith yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

voznesenskym committed:
Update on "Create nn_modules_sources structure, map params and placeholders"


This PR Introduces a new structure on the tracing context, a ModuleContext, with a mapping of name to source.

The original motivation is to (A) move dedup guards off positional arguments structures (B) Add support for dedup guards on params.

To that end, we lack a structure that is:

1) Not positional
2) Can record both guards and placeholders
3) Can help us get a source out for a given input or parameter

This PR introduces the structure mentioned above. The idea here is that, we can use this to keep a mapping of all param names, buffer names, and module names, as well as placeholder names, to their sources. Then, anywhere within the tracing context that has access to any of the above can always get their full, proper, source, and therefore, their name in user code. This is particularly important for creating guards in a system downstream of dynamo, where we do no have VariableTracker. 

cc soumith yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

voznesenskym committed:
Update on "Create nn_modules_sources structure, map params and placeholders"


This PR Introduces a new structure on the tracing context, a ModuleContext, with a mapping of name to source.

The original motivation is to (A) move dedup guards off positional arguments structures (B) Add support for dedup guards on params.

To that end, we lack a structure that is:

1) Not positional
2) Can record both guards and placeholders
3) Can help us get a source out for a given input or parameter

This PR introduces the structure mentioned above. The idea here is that, we can use this to keep a mapping of all param names, buffer names, and module names, as well as placeholder names, to their sources. Then, anywhere within the tracing context that has access to any of the above can always get their full, proper, source, and therefore, their name in user code. This is particularly important for creating guards in a system downstream of dynamo, where we do no have VariableTracker. 

cc soumith yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

voznesenskym committed:
Update on "Create nn_modules_sources structure, map params and placeholders"


This PR Introduces a new structure on the tracing context, a ModuleContext, with a mapping of name to source.

The original motivation is to (A) move dedup guards off positional arguments structures (B) Add support for dedup guards on params.

To that end, we lack a structure that is:

1) Not positional
2) Can record both guards and placeholders
3) Can help us get a source out for a given input or parameter

This PR introduces the structure mentioned above. The idea here is that, we can use this to keep a mapping of all param names, buffer names, and module names, as well as placeholder names, to their sources. Then, anywhere within the tracing context that has access to any of the above can always get their full, proper, source, and therefore, their name in user code. This is particularly important for creating guards in a system downstream of dynamo, where we do no have VariableTracker. 

cc soumith yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

voznesenskym committed:
Update on "Create nn_modules_sources structure, map params and placeholders"


This PR Introduces a new structure on the tracing context, a ModuleContext, with a mapping of name to source.

The original motivation is to (A) move dedup guards off positional arguments structures (B) Add support for dedup guards on params.

To that end, we lack a structure that is:

1) Not positional
2) Can record both guards and placeholders
3) Can help us get a source out for a given input or parameter

This PR introduces the structure mentioned above. The idea here is that, we can use this to keep a mapping of all param names, buffer names, and module names, as well as placeholder names, to their sources. Then, anywhere within the tracing context that has access to any of the above can always get their full, proper, source, and therefore, their name in user code. This is particularly important for creating guards in a system downstream of dynamo, where we do no have VariableTracker. 

cc soumith yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

voznesenskym committed:
Update on "Create nn_modules_sources structure, map params and placeholders"


This PR Introduces a new structure on the tracing context, a ModuleContext, with a mapping of name to source.

The original motivation is to (A) move dedup guards off positional arguments structures (B) Add support for dedup guards on params.

To that end, we lack a structure that is:

1) Not positional
2) Can record both guards and placeholders
3) Can help us get a source out for a given input or parameter

This PR introduces the structure mentioned above. The idea here is that, we can use this to keep a mapping of all param names, buffer names, and module names, as well as placeholder names, to their sources. Then, anywhere within the tracing context that has access to any of the above can always get their full, proper, source, and therefore, their name in user code. This is particularly important for creating guards in a system downstream of dynamo, where we do no have VariableTracker. 

cc soumith yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

ezyang commented on torch/_dynamo/output_graph.py:
```diff
@@ -441,27 +476,57 @@ def wrap_name(module_key):
         if not name or not name[0].isalpha():
             name = "sub" + name
         base = name
         for i in itertools.count():
```
I'm going to keep rejecting this PR until this loop no longer exists. ONE spot for name allocation, not two.

ezyang reviewed (changes requested):
see comment

voznesenskym committed:
Update on "Create nn_modules_sources structure, map params and placeholders"


This PR Introduces a new structure on the tracing context, a ModuleContext, with a mapping of name to source.

The original motivation is to (A) move dedup guards off positional arguments structures (B) Add support for dedup guards on params.

To that end, we lack a structure that is:

1) Not positional
2) Can record both guards and placeholders
3) Can help us get a source out for a given input or parameter

This PR introduces the structure mentioned above. The idea here is that, we can use this to keep a mapping of all param names, buffer names, and module names, as well as placeholder names, to their sources. Then, anywhere within the tracing context that has access to any of the above can always get their full, proper, source, and therefore, their name in user code. This is particularly important for creating guards in a system downstream of dynamo, where we do no have VariableTracker. 

cc soumith yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

voznesenskym commented:
Not high prio, wasting too much time on this
